### DISCUSSION TITLE
geo querys and sorting by distance

### DISCUSSION THREAD

**Discussion by Crey443** [supabase/supabase#5390] [Questions] [answered]
Hi, 
I am currently using firestore, but I need to be able to get my data via a geopoint that is stored in the documents.
Now I am wondering if supabase supports Geo queries and sorting by distance to a given point?

Thanks for any help.

**Comment by wiesson** [comment]
If your data is available as longitude and latitude (floats), then you could make use of haversine formula.

**Reply by Crey443** [reply]
when using this formula, is it possible to sort the data before getting it? 
because I want to paginate it for an infinite list view that is sorted by distance

**Reply by wiesson** [reply]
I'm not sure, but I guess yes - I have used it once with postgres (raw query), but not yet with supabase. So if you write your SQL query, check if you can add your sort parameters. Once it works, define a function and call with via RPC (others may know better how it works 😅 ) - https://supabase.com/docs/guides/database/functions and https://supabase.com/docs/reference/javascript/rpc

**Comment by GaryAustin1** [comment]
You can also use the postgres PostGIS extension with a point data type and then make queries (using the rpc functions mentioned above) for simple to complex retrieval of data.  It is reasonably complicated to get things going though, and you can't use the table UI to add the column.  But should be fast as built in and let you do  more advanced things like find in a shape versus just distance if ever needed.

**Comment by oripka** [comment]
I found this project https://www.madewithsupabase.com/p/supabase-geodb, maybe that's helpful

**Comment by dshukertjr** [comment, marked as answer]
Yes, Postgres supports geo queries much better than Firestore!

In order to query data and order it by distance, you would first need to enable [postgis extension](https://supabase.com/docs/guides/database/extensions) in your Supabase project.

Your table could look something like this:
```sql
create table if not exists public.videos (
    id uuid not null primary key DEFAULT uuid_generate_v4 (),
    created_at timestamp with time zone default timezone('utc' :: text, now()) not null,
    description varchar(320) not null,
    location geography(POINT) not null
);
```
The key here is the `location` column with `geography(POINT)` type which represents a geographic location on earth. 

Then you can create a Postgres function like this:
```sql
create or replace function test_nearby(lat float, long float)
returns setof videos
as $$
    select *
    from videos
    order by location <-> st_point(long, lat)::geography;
$$
language sql;
```

The above function can be called like this:

```typescript
 { data, error } = await supabase.rpc('nearby_videos', 
  {location: 'POINT(${longitude} ${latitude})'}
)
```

I actually utilize a lot of these geo queries in my open source project [Spot](https://github.com/dshukertjr/spot), so it might help out to check it out.

**Reply by phillipmohr** [reply]
@dshukertjr thanks for the solution! I have a collection of stores. Is it possible to get the 10 nearest stores with this way instead of just one?

**Reply by dshukertjr** [reply]
@phillipmohr 

Yup! Actually the query that I provided gets all of the locations(although Supabase automatically limits to 1000 by default) in nearest to farthest order. You can add `limit` to the `rpc` call on the client side code to limit the number of results you get back in a single call. 

```typescript
{ data, error } = await supabase.rpc('nearby_videos', 
  {location: 'POINT(${longitude} ${latitude})'}
)
.limit(10)
```

**Reply by phillipmohr** [reply]
oh, I didn't know this already returns multiple "stores". In my case, I only get 1 store returned and I'm not sure why...

Edit: I changed my function like this, [as mentioned here:](https://github.com/supabase/supabase/discussions/5668#discussioncomment-2254208)
```sql
create or replace function nearby_studios(location text)
returns setof studios
language sql
as
$$
  select *
  from studios
  order by location <-> st_geogfromtext(nearby_studios.location);
$$;
```

and now I get more than 1 result :)

**Reply by dshukertjr** [reply]
@phillipmohr 

I'm glad to hear that you got it working. I wonder what the cause of you not getting more than 1 result in your original one. Could you post the original SQL you had to create a function?

**Reply by phillipmohr** [reply]
@dshukertjr actually I just used your example and changed the variables.
```sql
create or replace function nearby_studios(location text)
returns studios
as
$func$
    select *
    from studios
    order by location <-> st_geogfromtext(nearby_studios.location);
$func$
language sql;
```

**Reply by dshukertjr** [reply]
@phillipmohr Thanks! I guess I learned something new about Postgres today! I have updated my answer for anyone seeing this in the future!

**Reply by dshukertjr** [reply]
And here is Alaister teaching me about it on a stream 😂
https://youtu.be/KWU_lyUEeZU?t=3727

**Reply by phillipmohr** [reply]
Haha 😂

**Reply by devry12** [reply]
I don't really understand postgre queries, I only understand sql queries, from the existing functions, how do I do that so that there is an additional distance between point a and point b

**Reply by dshukertjr** [reply]
@devry12 
Great question! 
You can use the [st_distance](https://postgis.net/docs/ST_Distance.html) function to get the distance between each row and the distance that you are calling from. `st_distance` will return the distance between two points in meters. 

You can edit the `nearby_videos` function like this to include the distance:
```sql
create or replace function nearby_videos(location text)
returns setof record
as
$$
    select id, name, st_distance(st_geogfromtext(nearby_videos.location), location) as distance
    from videos
    order by location <-> st_geogfromtext(nearby_videos.location);
$$
language sql;
```

**Reply by devry12** [reply]
@dshukertjr 
sorry, I had an error following the sql, if I tested it in the query editor it worked like sql but when I combined it with the function it got an error

"Failed to run sql query: return type mismatch in function declared"

**Reply by dshukertjr** [reply]
@devry12 Sorry, I had the wrong return type set. It was `setof videos`, but I have updated the above code to be `setof record`. Please try it again!

**Reply by devry12** [reply]
it works, thank you so much 😁😁

**Reply by phillipmohr** [reply]
@dshukertjr would you mind helping me with something else? 🙈 
I use this code at the moment:
```sql
create or replace function nearby_studios(location text)
returns setof studios
language sql
as
$$
  select *
  from studios
  order by location <-> st_geogfromtext(nearby_studios.location);
$$;
```

it works great, but there are also premium studios (stores) which I want to show first from all the results I get. Each studio has the column `isPremium` which is either `true` or `false`. I tried it like this:

Usage: 
```js
const { data: studios, error } = await client.rpc('nearby_studios',
    { location: `POINT(${userLocation.longitude} ${userLocation.latitude})` }
)
    .limit(50)
    .order('isPremium', { ascending: false })
```

But this seems to overwrite the geolocation sorting.

**Reply by dshukertjr** [reply]
@phillipmohr 
You could probably include the order by premium logic inside the function as well like this the following. Let me know if it works!

```sql
create or replace function nearby_studios(location text)
returns setof studios
language sql
as
$$
  select *
  from studios
  order by
    is_premium is true desc, 
    location <-> st_geogfromtext(nearby_studios.location);
$$;
```

**Reply by phillipmohr** [reply]
@dshukertjr thanks! Unfortunately I get `Failed to run sql query: column "is_premium" does not exist`
I also tried `isPremium` but that didn't work either.

Edit: I had to write it like this and not it works :) 
`"isPremium" is true desc, `

Thank you!

**Reply by dshukertjr** [reply]
@phillipmohr 
Sorry that I missed that one, but I'm glad to hear that you have figured it out!

In Postgres, when you have capital letter in your column name, you have to wrap them in `""`.  You can either make sure to wrap them with double quote every time, or an easier way would to just use lower snake case like `is_premium` as the naming convention for your columns!

**Reply by phillipmohr** [reply]
thanks for the explanation! I'm new to Postgres, so I didn't know :)

**Reply by phillipmohr** [reply]
@dshukertjr hey Tyler, Unfornuately it turned out that it doesn't work like I expected. This puts premium studios first, yes, but it also gives me premium studios from another city which is far away. I only want to get premium studios that are nearby. How could I do that?

**Edit:** Tyler helped me out somewhere else, this is the solution if anyone should stumble over this. This sorts by location and then puts premium studios first:
```sql
create or replace function nearby_studios(location text)
returns setof studios
language sql
as
$$
  with nearest_50_studios as (select *
    from studios
    order by
      location <-> st_geogfromtext(nearby_studios.location) limit 50
  )
  select * from nearest_50_studios order by is_premium is true desc, location <-> st_geogfromtext(nearby_studios.location);
$$;
```

**Reply by mijorus** [reply]
I edited it slightly to make it more cleaner to use, IMO

```sql
create or replace function get_nearby_shops(lat float, long float)
returns setof shops
language sql
as

$$
  select *
  from shops
  order by location <-> st_geogfromtext(
    concat('POINT(', get_nearby_shops.long, ' ', get_nearby_shops.lat, ')')
);
$$;
```

```js
const {data, err} = await client.rpc('get_nearby_shops', {lat: 30, long: 20}).limit(10)
```

I have also found another solution:

```sql
SELECT uuid, name, lat, long, (point(long, lat) <@> point(get_nearby_shops.long, get_nearby_shops.lat)) AS distance
FROM
 shops

ORDER BY
 distance;
```

although some claim it is slower on bigger datasets

**Reply by organicnz** [reply]
Hi @dshukertjr, 

Apologies for asking, mate. Just wondering, Is it possible to convert geography to geopoint data type? I already switched from Firebase to Supabase where I used lat/long geopoint data type in my FlutterFlow application, but now I can't set it up unfortunately :)

Thank you :)

<img width="1440" alt="Screenshot_2023-02-12_at_1_43_51_AM" src="https://user-images.githubusercontent.com/4765090/218284258-e25e4d63-8c14-4579-a6ec-63d56c94d7e4.png">

**Reply by dshukertjr** [reply]
@organicnz 
No need to apologize to ask questions 😁

The FlutterFlow Supabase integration is build and maintained by the FlutterFlow team, and we don't have access to any of their source codes. These data types seem FlutterFlow specific, so you would have to ask the FlutterFlow team how to convert the Supabase geography type to data types that you can use within FlutterFlow.

**Reply by incrediblezayed** [reply]
Hey @diamondalltheway @dshukertjr @phillipmohr @mijorus Can you guys help me to filter by distance, for example I want to fetch the stores within 50 Kms/Miles of the given lat long, how can I achieve this?

**Comment by iamevaristus** [comment]
I don't really know how, but I want to build something like uber where the driver's location will be inserted to the live_location table so that the users can fetch that when they want to request for a ride. I have been able to insert the driver's location to the table, but how can I make the table realtime such that it reads the changes of the driver's location and provide the new latlng. Like when the driver moves, I need the table to be updated with the new latlng.

I am new to supabase and I need help with that. I dont want to be adding new data row to my live_location table, i want to update a certain row based on the ID of that driver such that a driver will only have one row of data which will be inserted once and then be updating itself once the location changes. And when the driver goes offline, I want to delete that record from the live_location table.
